src/interfaces/ffi/api.rs

```diff
@@ -396,3 +300,295 @@ pub unsafe extern "C" fn h_compact(
+/// - `label`                   : additional information used to derive Entry
+///   Table UIDs
+/// - `keywords`                : `serde` serialized list of base64 keywords
+/// - `max_results_per_keyword` : maximum number of results returned per keyword
+/// - `max_depth`               : maximum recursion depth allowed
+/// - `progress_callback`       : callback used to retrieve intermediate results
+///   and transmit user interrupt
+/// - `fetch_entry`             : callback used to fetch the Entry Table
+/// - `fetch_chain`             : callback used to fetch the Chain Table
+///
+/// # Safety
+///
+/// Cannot be safe since using FFI.
+pub unsafe extern "C" fn h_search_cloud(
+    search_results_ptr: *mut c_char,
+    search_results_len: *mut c_int,
+    token_ptr: *const c_char,
+    label_ptr: *const u8,
+    label_len: c_int,
+    keywords_ptr: *const c_char,
+    max_results_per_keyword: c_int,
+    max_depth: c_int,
+    fetch_chains_batch_size: c_uint,
+) -> c_int {
+    let token = ffi_read_string!("keywords", token_ptr);
+
+    let findex = ffi_unwrap!(FindexCloud::new(token, None));
+    let master_key = findex.token.findex_master_key.clone();
+
+    ffi_search(
+        findex,
+        &master_key,
+        search_results_ptr,
+        search_results_len,
+        label_ptr,
+        label_len,
+        keywords_ptr,
+        max_results_per_keyword,
+        max_depth,
+        fetch_chains_batch_size,
+    )
+}
+
+#[no_mangle]
+/// Index the given values for the given keywords. After upserting, any
+/// search for such a keyword will result in finding (at least) the
+/// corresponding value.
+///
+/// # Serialization
+///
+/// The list of values to index for the associated keywords should be serialized
+/// as follows:
+///
+/// `LEB128(n_values) || serialized_value_1
+///     || LEB128(n_associated_keywords) || serialized_keyword_1 || ...`
+///
+/// where values serialized as follows:
+///
+/// `LEB128(value_bytes.len() + 1) || base64(prefix || value_bytes)`
+///
+/// with `prefix` being `l` for a `Location` and `w` for a `NextKeyword`, and
+/// where keywords are serialized as follows:
+///
+/// `LEB128(keyword_bytes.len()) || base64(keyword_bytes)`
+///
+/// # Parameters
+///
+/// - `token`           : Findex Cloud token
+/// - `label`           : additional information used to derive Entry Table UIDs
+/// - `indexed_values_and_keywords` : serialized list of values and the keywords
+///   used to index them
+///
+/// # Safety
+///
+/// Cannot be safe since using FFI.
+pub unsafe extern "C" fn h_upsert_cloud(
+    token_ptr: *const c_char,
+    label_ptr: *const u8,
+    label_len: c_int,
+    indexed_values_and_keywords_ptr: *const c_char,
+) -> c_int {
+    let token = ffi_read_string!("keywords", token_ptr);
+
+    let findex = ffi_unwrap!(FindexCloud::new(token, None));
+    let master_key = findex.token.findex_master_key.clone();
+
+    ffi_upsert(
+        findex,
+        &master_key,
```

Comment:
Why do you need to clone this variable?
```suggestion
    ffi_upsert(
        findex,
        &findex.token.findex_master_key,
```

Reply:
Because `findex` is move inside the 1rst argument so Rust doesn't allow this… Not a big deal right now, but could be improved for sure at another time.

Reply:
True, I didn't see it. Why do you need the token to be part of Findex?

Reply:
Because the token contains the keys and the keys are required in the callbacks.